File: sdks/javascript/src/hash.ts
import { createHash } from 'node:crypto'

/**
 * Returns a deterministic bucket (0-99) for a given flag key and user ID.
 * Uses SHA-256 for distribution.
 *
 * MUST match the Go backend: binary.BigEndian.Uint64(sha256(flagKey+userID)[:8]) % 100
 */
export function consistentHash(flagKey: string, userId: string): number {
  const hash = createHash('sha256').update(flagKey + userId).digest()
  // Read first 8 bytes as big-endian uint64 — matches Go backend
  const value = hash.readBigUInt64BE(0)
  return Number(value % 100n)
}

File: sdks/go/hash.go
package togglerino

import (
	"crypto/sha256"
	"encoding/binary"
)

// consistentHash returns a deterministic bucket (0-99) for a given flag key
// and user ID. Uses SHA-256 for distribution. Must match the backend's
// ConsistentHash in internal/evaluation/hash.go exactly.
func consistentHash(flagKey, userID string) int {
	h := sha256.Sum256([]byte(flagKey + userID))
	// Take first 8 bytes as big-endian uint64.
	n := binary.BigEndian.Uint64(h[:8])
	return int(n % 100)
}
